WIP: add gas fraction phase diagrams and VR plots

wmfw23 opened the pull request:
*Number of changes to provide Hydrogen Species plots:*

- Added template VR config file: `colibre/vrconfig_example.cfg` 
- Added local universe observational species fractions comparison plots  H2 (Saintonge+17) and HI (Catinella+18)
- Added HI and H2 mass fractions for local Universe
- Added phase plots coloured by species fractions

*Notes*:

This is designed for COLIBRE runs using `--with-cooling=CHIMES` , as the VR config file requires knowledge of which H species are assigned to which column. It would be better if this was universal for cooling modes, but CHIMES is the current default so have written it for that case. This update takes advantage of recent updates to the pipeline script to allow argument feeding.

Marking as WIP, in case there are better ways to integrate in the VR configuration. This also foreshadows issues with upcoming dust plots which should be stable for different dust models.

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+setattr(self, f"total_dust_masses", total_dust_mass)
+
+#species fraction properties
+gas_mass = catalogue.masses.m_gas
```
Can we have some documentation here about why we're using a different gas mass for the species fractions and dust fractions?

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+
+#species fraction properties
+gas_mass = catalogue.masses.m_gas
+H_frac = catalogue.element_mass_fractions.element_0
```
This section needs protecting for cases where the fractions are not present (see above with the `try`, `except`)

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+#species fraction properties
+gas_mass = catalogue.masses.m_gas
+H_frac = catalogue.element_mass_fractions.element_0
+gal_area = np.pi * catalogue.projected_apertures.projected_2_rhalfmass_star_30_kpc**2
```
```
galaxy_radii = catalogue.projected_apertures.projected_1_rhalfmass_star_30_kpc
galaxy_area = np.pi * (galaxy_radii * galaxy_radii)
```
We should use aperture 1 here as that's what used for the galaxy size plot in the pipeline.

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+gas_mass = catalogue.masses.m_gas
+H_frac = catalogue.element_mass_fractions.element_0
+gal_area = np.pi * catalogue.projected_apertures.projected_2_rhalfmass_star_30_kpc**2
+mstar_30 = catalogue.projected_apertures.projected_2_mass_star_30_kpc.to('Msun')
```
The unit conversion here is not necessary, see below.

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+mstar_30 = catalogue.projected_apertures.projected_2_mass_star_30_kpc.to('Msun')
+
+# NB: this is designed for CHIMES species arrays
+self.neutral_hydrogen_masses = gas_mass * H_frac * catalogue.species_fractions.species_1
```
Oddly the velociraptor names are singular, so this should be `neutral_hydrogen_mass` (etc.)

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+self.neutral_hydrogen_masses = gas_mass * H_frac * catalogue.species_fractions.species_1
+self.ionized_hydrogen_masses = gas_mass * H_frac * catalogue.species_fractions.species_2
+self.molecular_hydrogen_masses =  gas_mass * H_frac * catalogue.species_fractions.species_7
+self.hi_to_stellar_mass = self.neutral_hydrogen_masses / catalogue.masses.m_star
```
Please include aperture name here.

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+self.h2_to_stellar_mass = self.molecular_hydrogen_masses / catalogue.masses.m_star
+self.mu_star = mstar_30 / gal_area
+
+self.xgass_sel = (mstar_30 > (1e9 * mstar_30.units)) * (mstar_30 < (pow(10,11.5) * mstar_30.units))
```
Can we have a more descriptive name here? I'm not sure what this means. Please also use `logical_and` instead of multiplying.

```
self.{new_name} = np.logical_and(
    catalogue.apertures.mass_star_30_kpc > unyt.unyt_quantity(10**9, "Solar_Mass"),
    catalogue.apertures.mass_star_30_kpc < unyt.unyt_quantity(10**(11.5), "Solar_Mass"),
)
```

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+self.mu_star = mstar_30 / gal_area
+
+self.xgass_sel = (mstar_30 > (1e9 * mstar_30.units)) * (mstar_30 < (pow(10,11.5) * mstar_30.units))
+self.xcoldgass_sel = (mstar_30 > (1e9 * mstar_30.units)) * (mstar_30 < (pow(10,11.5) * mstar_30.units))
```
Same comment as above. This also is the same line of code, unless I'm mistaken - should these not be different?

wmfw23 replied:
These are the same for now, but technically represent different surveys, so keeping as distinct selection functions

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+self.xgass_sel = (mstar_30 > (1e9 * mstar_30.units)) * (mstar_30 < (pow(10,11.5) * mstar_30.units))
+self.xcoldgass_sel = (mstar_30 > (1e9 * mstar_30.units)) * (mstar_30 < (pow(10,11.5) * mstar_30.units))
+
+self.neutral_hydrogen_masses.name = "HI Mass"
```
Please include aperture here

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+self.xcoldgass_sel = (mstar_30 > (1e9 * mstar_30.units)) * (mstar_30 < (pow(10,11.5) * mstar_30.units))
+
+self.neutral_hydrogen_masses.name = "HI Mass"
+self.molecular_hydrogen_masses.name = "H2 Mass"
```
I suppose we can include LaTeX formatting here, so "H$_2$ Mass"

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+self.ionized_hydrogen_masses.name = "HII Mass"
+
+self.hi_to_stellar_mass.name = 'HI to Stellar Mass Fraction'
+self.h2_to_stellar_mass.name = 'H2 to Stellar Mass Fraction'
```
H$_2$ again. Also, please ensure you use double quotes throughout and include the aperture.

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+self.hi_to_stellar_mass.name = 'HI to Stellar Mass Fraction'
+self.h2_to_stellar_mass.name = 'H2 to Stellar Mass Fraction'
+
+self.mu_star.name = f"Stellar Surface Density"
```
Aperture would be useful again or it could be nice to have `$\pi R_{*, 30 {\rm kpc}}^2 / M_{*, 30 {\rm kpc}}$` here too to remove any ambiguity.

JBorrow commented on colibre/auto_plotter/gas_fraction_plots.yml:
```diff
@@ -0,0 +1,183 @@
+      value: 4e11
+      units: Solar_Mass
+  metadata:
+    title: Stellar Mass-H2 Fraction
```
Please include a caption, and it's worth having the H$_2$ here.

JBorrow commented on colibre/auto_plotter/gas_fraction_plots.yml:
```diff
@@ -0,0 +1,183 @@
+      units: "1 / Gigayear"
+  metadata:
+    title: Stellar Mass-H2 Fraction
+    section: H_2 Fractions
```
Same commen  as above.

JBorrow commented on colibre/auto_plotter/gas_fraction_plots.yml:
```diff
@@ -0,0 +1,183 @@
+      value: 1e4
+      units: "1 / Gigayear"
+  metadata:
+    title: Stellar Mass-H2 Fraction
```
Isn't this plot `SFR-H$_2$ Fraction`?

JBorrow commented on colibre/auto_plotter/gas_fraction_plots.yml:
```diff
@@ -0,0 +1,183 @@
+      value: 1e11
+      units: "Solar_Mass / (kpc**2)"
+  metadata:
+    title: Stellar Mass Surface Density-H2 Fraction
```
Caption again.

JBorrow commented on colibre/auto_plotter/gas_fraction_plots.yml:
```diff
@@ -0,0 +1,183 @@
+  median:
+    plot: true
+    log: true
+    number_of_bins: 9
```
I don't fully understand why we would want an odd number of bins here. Is it worth expanding the plot limits to 1e9 -> 1e12, and have there be 15 equal-width bins of 0.2 dex here, consistent with the rest of the plots? We can then use the `adaptive: true` to use adaptive binning too.

JBorrow commented on colibre/auto_plotter/gas_fraction_plots.yml:
```diff
@@ -0,0 +1,183 @@
+  median:
+    plot: true
+    log: true
+    number_of_bins: 29
```
This should be 40 to be consistent with the rest using 0.2 dex bins. Also worth using `adaptive: true` everywhere.

JBorrow commented on colibre/auto_plotter/gas_fraction_plots.yml:
```diff
@@ -0,0 +1,183 @@
+  median:
+    plot: true
+    log: true
+    number_of_bins: 9
```
Same binning comment as previously. Here it would be 25 bins.

JBorrow commented on colibre/auto_plotter/gas_fraction_plots.yml:
```diff
@@ -0,0 +1,183 @@
+      value: 1e4
+      units: "1 / Gigayear"
+  metadata:
+    title: Stellar Mass-HI Fraction
```
This plot shows sSFR not stellar mass

JBorrow commented on colibre/auto_plotter/gas_mass_function.yml:
```diff
@@ -0,0 +1,37 @@
+hi_mass_function:
```
Can we have these be adaptive mass functions? `type: "adaptivemassfunction"`

JBorrow commented on colibre/auto_plotter/gas_mass_function.yml:
```diff
@@ -0,0 +1,37 @@
+hi_mass_function:
+  type: "massfunction"
+  legend_loc: "lower left"
+  number_of_bins: 25
```
The number of bins should be set to be 0.2 dex in mass.

JBorrow commented on colibre/scripts/density_temperature_species.py:
```diff
@@ -0,0 +1,191 @@
+# Set the limits of the figure.
+density_bounds = [10 ** (-9.5), 1e6]  # in nh/cm^3
+temperature_bounds = [10 ** (0), 10 ** (9.5)]  # in K
+dustfracs_bounds = [-2, 0]  # In metal mass fraction
```
Is it possible to specify these in linear space (`dustfracs_bounds = [1e-2, 1e0]`), and use that throughout with  logarithmic normalizations instead of linear ones on log quantities? That's what we do in all other cases so it would be nice here too.

JBorrow reviewed (commented):
A number of comments:

1. Let's make all of the plots `adaptive`, we might as well.
2. Let's make sure that the pipeline runs even with missing data (i.e. that `registration.py` doesn't crash on missing data).

More detailed comments in the review.

JBorrow commented:
You may also need to rebase.

wmfw23 commented on colibre/registration.py:
```diff
@@ -118,3 +118,44 @@
+
+# iterate through available dust types
+dust_idx = 0
+total_dust_fraction = metal_mass_fraction_gas*0.
```
Thanks for the pointers, quick question: would you prefer to have `total_dust_mass` undefined in the no explicit dust case, or to define it as a zero array so that dust scripts still run without crashing?

JBorrow replied:
Let's have it defined as a zero array but change the name to state that it wasn't found.

wmfw23 commented:
Thanks Josh, I've incorporated your comments where possible and rebased. There's not a great answer for which aperture to use, so using 100 kpc aperture for gas and dust fractions for now, as 30 kpc + are largely equivalent for the simulations volumes running for COLIBRE, and 100 kpc should more or less capture al galaxy material. I've opened an issue on my branch to look at this further but i think for now this works

I noticed some problems, e.g. the `adaptivemassfunction` type doesn't seem to work for arrays where the whole catalogue is out-of-bounds, so reverting to regular `massfunction` for now.

Also noticed formatting with `black` reformats lots of files that I haven't touched, so may want to do this on the master branch (not committing those files to avoid confusing)

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,85 @@
 
     setattr(self, f"stellar_mass_to_halo_mass_{aperture_size}_kpc", smhm)
+
+# if present iterate through available dust types
+
+
+# print(getattr(catalogue.dust_mass_fractions, 'dust_0'))
+
+# print([getattr(catalogue.dust_mass_fractions, sub_path) for sub_path in dir(catalogue.dust_mass_fractions) if sub_path.startswith('dust_')])
+
+print("done")
```
Remove prints and these comments please

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,77 @@
+)
+mstar_100 = catalogue.projected_apertures.projected_1_mass_star_100_kpc
+
+self.xgass_galaxy_selection = np.logical_and(
+    catalogue.apertures.mass_star_100_kpc > unyt.unyt_quantity(10 ** 9, "Solar_Mass"),
+    catalogue.apertures.mass_star_100_kpc
+    < unyt.unyt_quantity(10 ** (11.5), "Solar_Mass"),
+)
+
+self.xcoldgass_galaxy_selection = np.logical_and(
+    catalogue.apertures.mass_star_100_kpc > unyt.unyt_quantity(10 ** 9, "Solar_Mass"),
+    catalogue.apertures.mass_star_100_kpc
+    < unyt.unyt_quantity(10 ** (11.5), "Solar_Mass"),
+)
```
These still appear to be the same thing - and I have no idea what they are. Can we please have more descriptive names and a reason for why this must be repeated?

wmfw23 commented:
As I mentioned, these are selection functions mimicking the XGASS and XCOLDGASS survey mass selection. These are the same selection function but correspond to different surveys and published works. That's why I've separated them, despite being the same. If we want to add more specific selection criteria for either later, we can amend these. I'll add some comments to this effect.

JBorrow commented on colibre/registration.py:
```diff
@@ -118,3 +118,77 @@
+self.mu_star = mstar_100 / gal_area
+self.mu_star.name = "$\\pi R_{*, 100 {\\rm kpc}}^2 / M_{*, 100 {\\rm kpc}}$"
+
+self.neutral_hydrogen_mass = catalogue.masses.m_star * 0.0
+self.molecular_hydrogen_mass = catalogue.masses.m_star * 0.0
+self.h2_to_stellar_mass = catalogue.apertures.zmet_star_100_kpc * 0.0
+self.hi_to_stellar_mass = catalogue.apertures.zmet_star_100_kpc * 0.0
+
+self.neutral_hydrogen_mass.name = "HI Mass (100 kpc)"
+self.hi_to_stellar_mass.name = "HI to Stellar Mass Fraction (100 kpc)"
+self.molecular_hydrogen_mass.name = "H$_2$ Mass (100 kpc)"
+self.h2_to_stellar_mass.name = "H$_2$ to Stellar Mass Fraction (100 kpc)"
+
+try:
+    H_frac = catalogue.element_mass_fractions.element_0
+    # NB: this is designed for CHIMES species arrays
+    try:
+        self.neutral_hydrogen_mass = (
+            gas_mass * H_frac * catalogue.species_fractions.species_1
+        )
+        self.hi_to_stellar_mass = self.neutral_hydrogen_mass / catalogue.masses.m_star
+    except AttributeError:
+        self.neutral_hydrogen_mass.name += " not found (no species field)"
+        self.hi_to_stellar_mass.name += " not calculable (no species field)"
+    try:
+        self.molecular_hydrogen_mass = (
+            gas_mass * H_frac * catalogue.species_fractions.species_7
+        )
+        self.h2_to_stellar_mass = self.molecular_hydrogen_mass / catalogue.masses.m_star
+    except AttributeError:
+        self.molecular_hydrogen_mass.name += " not found (no species field)"
+        self.h2_to_stellar_mass.name += " not calculable (no species field)"
+
+except AttributeError:
+    self.molecular_hydrogen_mass.name += " not found (no H abundance)"
+    self.neutral_hydrogen_mass.name += " not found (no H abundance)"
+    self.hi_to_stellar_mass.name += " not calculable (no H abundance)"
+    self.h2_to_stellar_mass.name += " not calculable (no H abundance)"
```
There are a number of things here:

H2 to stellar mass etc. doesn't use the 100 kpc aperture stellar mass (that is apertures.mass_star_100_kpc). Please also include aperture size in these names.

Alongside this, the code path is a bit messy - I don't think we need things to be nested and I like to avoid string concatenation with +s wherever possible.

```
try:
   H_frac = catalogue.element_mass_fractions.element_0
   hydrogen_frac_error = ""
except:
   H_frac = 0.0
   hydrogen_frac_error = "(no H abundance)"
   
   
# Test for CHIMES arrays
try:
   species_1 = catalogue.species_fractions.species_1
   species_7 = catalogue.species_fractions.species_7
   species_frac_error = ""
except:
   species_1 = 0.0
   species_7 = 0.0
   species_frac_error = "(no species field)"
   
total_error = f" {species_frac_error} {hydrogen_frac_error}"

self.neutral_hydrogen_mass = gas_mass * H_frac * species_1
self.hi_mass_to_stellar_mass_100_kpc = (
   self.neutral_hydrogen_mass / catalogue.apertures.mass_star_100_kpc
)
self.molecular_hydrogen_mass = gas_mass * H_frac * species_7
self.h2_mass_to_stellar_mass_100_kpc = (
   self.molecular_hydrogen_mass / catalogue.apertures.mass_star_100_kpc
)

self.neutral_hydrogen_mass.name = f"HI Mass (100 kpc) {total_error}"
self.hi_to_stellar_mass.name = f"HI to Stellar Mass Fraction (100 kpc) {total_error}"
self.molecular_hydrogen_mass.name = f"H$_2$ Mass (100 kpc) {total_error}"
self.h2_to_stellar_mass.name = f"H$_2$ to Stellar Mass Fraction (100 kpc) {total_error}"
```

wmfw23 commented:
applied these changes and propagated through scripts. Updated all registered quantity variable names to refer to aperture where relevant

JBorrow commented:
Excellent! Thanks again for your contribution.
